### DISCUSSION TITLE
DBus and swift

### DISCUSSION THREAD

**Discussion by soubiliz** [AparokshaUI/Adwaita#39] [General]
DBus is a quite important building block of modern gtk apps, and for now I have found only a old and unmaintened module for it with swift : https://github.com/PureSwift/DBus

**Comment by david-swift** [comment]
Thank you for bringing in this aspect! I first want to focus on the UI itself, but I'll definitely look into this at some point. It is possible to directly access the C API with Swift's C interoperability features at the moment (although not very nice).

**Reply by soubiliz** [reply]
Most apps avoid using libdbus which is really low level, I myself add to work on dbus for vlc, and instead decided to use sd-bus or it's remplimentation (basu, and elogind?)

GUI apps can also use qdbus and glib implémentation.

But I think the best for us would be to to what zbus do in rust but for swift 

https://github.com/dbus2/zbus

**Reply by david-swift** [reply]
Thanks for the additional information. The API design of zbus looks similar to many Swift APIs (using property wrappers or macros). But as I have never used DBus before, this is probably too ambitious for me. Maybe it would be possible to provide a similar interface using GDBus?

**Reply by soubiliz** [reply]
Maybe zbus style would be nice if we want something that can be used outside of just gnome/GTK, but I don't see any issue with using GDBus I guess?

I haven't coded anything in swift yet and want to learn but for a first projects it will be a bit much to tackle.

A GDBus wrapper would be a good first step

**Comment by david-swift** [comment]
This discussion has moved to https://forums.aparoksha.dev/d/3-d-bus-for-swift
